src/cpu_feature/arm_linux/arm_linux.rs

```diff
@@ -0,0 +1,514 @@
+
+/// returns a u32 with bits set for use in GFp_armcap_P
+#[cfg(any(test, all(any(target_arch = "arm", target_arch = "aarch64"),
+    target_os="linux")))]
+fn armcap_bits<G: GetauxvalProvider> (cpuinfo: &CpuInfo,
+                                      procfs_auxv: &AuxVals<AuxvUnsignedLongNative>,
+                                      auxval_types: AuxvTypes<AuxvUnsignedLongNative>,
+                                      hwcap_features: ArmHwcapFeatures<AuxvUnsignedLongNative>,
+                                      getauxval_provider: G)
+        -> u32 {
+    let mut hwcap = AuxvUnsignedLongNative::from(0_u32);
+
+    // |getauxval| is not available on Android until API level 20. If it is
+    // unavailable, read from /proc/self/auxv as a fallback. This is unreadable
+    // on some versions of Android, so further fall back to /proc/cpuinfo.
+    // See
+    // https://android.googlesource.com/platform/ndk/+/882ac8f3392858991a0e1af33b4b7387ec856bd2
+    // and b/13679666 (Google-internal) for details. */
+
+    if let Some(v) = getauxval_provider.getauxval(auxval_types.AT_HWCAP) {
+        hwcap = v;
+    } else if let Some(v) = procfs_auxv.get(&auxval_types.AT_HWCAP) {
+        hwcap = *v;
+    } else if let Some(v) = hwcap_from_cpuinfo(&cpuinfo, hwcap_features) {
+        hwcap = v;
+    }
+
+    // Clear NEON support if known broken
+    if cpu_has_broken_neon(&cpuinfo) {
+        hwcap &= !hwcap_features.ARM_HWCAP_NEON;
+    }
+
+    // Matching OpenSSL, only report other features if NEON is present
+    let mut armcap: u32 = 0;
+    if hwcap & hwcap_features.ARM_HWCAP_NEON > AuxvUnsignedLongNative::from(0_u32) {
+        armcap |= ARMV7_NEON;
+
+        // Some ARMv8 Android devices don't expose AT_HWCAP2. Fall back to
+        // /proc/cpuinfo. See https://crbug.com/596156
+
+        let mut hwcap2 = AuxvUnsignedLongNative::from(0_u32);
+        if let Some(v) = getauxval_provider.getauxval(auxval_types.AT_HWCAP2) {
+            hwcap2 = v;
+        } else if let Some(v) = procfs_auxv.get(&auxval_types.AT_HWCAP2) {
+            hwcap2 = *v;
+        } else if let Some(v) = hwcap2_from_cpuinfo(&cpuinfo, hwcap_features) {
+            hwcap2 = v;
+        }
+
+        armcap |= armcap_for_hwcap2(hwcap2, hwcap_features);
+    }
+
+    return armcap;
+}
+
+fn armcap_for_hwcap2<T: AuxvUnsignedLong>(hwcap2: T,
+                                          hwcap_features: ArmHwcapFeatures<T>)
+                                          -> u32 {
+    let mut ret: u32 = 0;
+    if hwcap2 & hwcap_features.ARM_HWCAP2_AES > T::from(0) {
+        ret |= ARMV8_AES;
+    }
+    if hwcap2 & hwcap_features.ARM_HWCAP2_PMULL > T::from(0) {
+        ret |= ARMV8_PMULL;
+    }
+    if hwcap2 & hwcap_features.ARM_HWCAP2_SHA1 > T::from(0) {
+        ret |= ARMV8_SHA1;
+    }
+    if hwcap2 & hwcap_features.ARM_HWCAP2_SHA2 > T::from(0) {
+        ret |= ARMV8_SHA256;
+    }
+
+    return ret;
+}
+
+fn hwcap_from_cpuinfo<T: AuxvUnsignedLong>(cpuinfo: &CpuInfo,
```

Comment:
Depends on what you expect as a reader from `cpuinfo`, I guess -- I kept that module completely hardware agnostic. At least in its current form, it only exists to put a pretty face on the format of `/proc/cpuinfo` (just as auxv.rs is only a nice way to access the aux vector), and all arm-specific decisions based on that data are elsewhere. For context, my instinct with this sort of thing is to try to make things general purpose so as to avoid leaking concerns across layers unless there are performance concerns or something like that that would preclude such a design. As it is, you could hoist cpuinfo.rs pretty much as is into its own crate and it wouldn't leak any of its arm-feature-detection heritage, and that feels right to me, but I'm not tied to that structure if it bugs you.